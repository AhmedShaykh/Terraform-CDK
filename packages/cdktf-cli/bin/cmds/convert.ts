import yargs from "yargs";
import { convert } from "@cdktf/hcl2cdk";
import { displayVersionMessage } from "./version-check";

function readStreamAsString(stream: typeof process.stdin): Promise<string> {
  return new Promise((ok, ko) => {
    if (stream.isTTY) {
      ko(
        "No stdin was passed, please use it like this: cat main.tf | cdktf convert > imported.ts"
      );
    } else {
      let string = "";
      stream.on("data", (data) => (string += data.toString()));

      stream.on("close", () => ok(string));
      stream.on("error", (err) => ko(err));
    }
  });
}

class Command implements yargs.CommandModule {
  public readonly command = "convert [OPTIONS]";
  public readonly describe =
    "Converts a single file of HCL configuration to Terraform CDK. Takes the file to be converted on stdin.";

  public readonly builder = (args: yargs.Argv) =>
    args
      .example(
        "cat main.tf | cdktf convert",
        "Takes the HCL content of main.tf and converts it to Terraform CDK content and prints it"
      )
      .example(
        "cat main.tf | cdktf convert > imported.ts",
        "Takes the HCL content of main.tf and converts it to Terraform CDK content in imported.ts"
      )
      .example(
        "cat main.tf | cdktf convert --language python > imported.py",
        "Takes the HCL content of main.tf and converts it to Terraform CDK content in imported.ts"
      )
      .option("language", {
        choices: ["typescript", "python", "csharp", "java"],
        default: "typescript",
      })
      .showHelpOnFail(true);

  public async handler({ language }: any) {
    await displayVersionMessage();

    const input = await readStreamAsString(process.stdin);
    const { all: output } = await convert(input, {
      language,
    });
    console.log(output);
  }
}

module.exports = new Command();
