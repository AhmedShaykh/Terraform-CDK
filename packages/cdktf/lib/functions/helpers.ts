// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Tokenization, Token } from "../tokens/token";
import { call } from "../tfExpression";
import { IResolvable } from "../tokens/resolvable";
import { TokenString, extractTokenDouble } from "../tokens/private/encoding";

// We use branding here to ensure we internally only handle validated values
// this allows us to catch usage errors before terraform does in some cases
type TFValue = any & { __type: "tfvalue" };
type TFValueValidator = (value: any) => TFValue;

type ExecutableTfFunction = (...args: any[]) => IResolvable;

/**
 * Determines if given str has unescaped double quotes
 * @param str String to test
 * @returns Boolean
 */
function hasUnescapedDoubleQuotes(str: string) {
  return /(^|[^\\])([\\]{2})*"/.test(str);
}

// Validators
// eslint-disable-next-line jsdoc/require-jsdoc
export function anyValue(value: any): any {
  return value;
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function mapValue(value: any): any {
  return value;
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function stringValue(value: any): any {
  if (typeof value !== "string" && !Tokenization.isResolvable(value)) {
    throw new Error(`'${value}' is not a valid string nor a token`);
  }

  if (typeof value === "string" && hasUnescapedDoubleQuotes(value)) {
    throw new Error(
      `'${value}' can not be used as value directly since it has unescaped double quotes in it. To safely use the value please use Fn.rawString on your string.`
    );
  }

  return value;
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function numericValue(value: any): any {
  if (typeof value !== "number" && !Tokenization.isResolvable(value)) {
    throw new Error(`${value} is not a valid number nor a token`);
  }
  return value;
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function listOf(type: TFValueValidator): TFValueValidator {
  return (value: any) => {
    if (Tokenization.isResolvable(value)) {
      return value;
    }

    if (!Array.isArray(value)) {
      //   throw new Error(`${value} is not a valid list`);
      return value;
    }

    return value
      .filter((item) => item !== undefined && item !== null)
      .map((item, i) => {
        if (Tokenization.isResolvable(item)) {
          return item;
        }

        if (TokenString.forListToken(item).test()) {
          return item;
        }

        if (extractTokenDouble(item, true) !== undefined) {
          return item;
        }

        if (TokenString.forMapToken(item).test()) {
          return item;
        }

        if (typeof item === "string") {
          const tokenList = Tokenization.reverseString(item);
          const numberOfTokens =
            tokenList.tokens.length + tokenList.intrinsic.length;
          if (numberOfTokens === 1 && tokenList.literals.length === 0) {
            return item;
          }
        }

        try {
          type(item);
          return typeof item === "string" ? `"${item}"` : item;
        } catch (error) {
          throw new Error(
            `Element in list ${value} at position ${i} is not of the right type: ${error}`
          );
        }
      });
  };
}

// Tokenization
// eslint-disable-next-line jsdoc/require-jsdoc
export function asString(value: IResolvable) {
  return Token.asString(value);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asNumber(value: IResolvable) {
  return Token.asNumber(value);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asList(value: IResolvable) {
  return Token.asList(value);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asStringMap(value: IResolvable) {
  return Token.asStringMap(value);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asBoolean(value: IResolvable) {
  return value; // Booleans can not be represented as a token
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asAny(value: IResolvable) {
  // Ordinarily casting to any can cause issues, but
  // in this case it makes using functions a bit easier in TS
  // and doesn't really harm other languages.
  // Jsii has issues when returning the value directly,
  // so wrap as a string.
  return asString(value) as any;
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function terraformFunction(
  name: string,
  argValidators: TFValueValidator | TFValueValidator[]
): ExecutableTfFunction {
  return function (...args: any[]) {
    if (Array.isArray(argValidators)) {
      if (args.length !== argValidators.length) {
        throw new Error(
          `${name} takes ${argValidators.length} arguments, but ${args.length} were provided`
        );
      }
      return call(
        name,
        args.map((arg, i) => {
          try {
            return argValidators[i](arg);
          } catch (error) {
            throw new Error(
              `Argument ${i} of ${name} failed the validation: ${error}`
            );
          }
        })
      );
    } else {
      return call(name, argValidators(args));
    }
  };
}
