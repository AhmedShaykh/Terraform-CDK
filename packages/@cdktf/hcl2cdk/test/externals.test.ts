import { testCase, Synth, binding } from "./helpers/convert";

describe("externals", () => {
  testCase.test(
    "handle external data",
    `
    provider "aws" {
      region = "us-east-1"
    }
    
    provider "external" {
    }

    resource "aws_eks_cluster" "changeme_spot_and_fargate_eks_cluster" {
      name = "foo"
      role_arn = "arn:aws:iam::123456789012:role/eksctl-foo-cluster-ServiceRole-1A2B3C4D5E6F"
      vpc_config {
        subnet_ids = ["subnet-12345678901234567"]
      }
    }

    data "aws_eks_cluster" "changeme_spot_and_fargate_eks_cluster_data" {
      name = "foo"
    }

    variable "changeme_spot_and_fargate_region" {
      type    = string
      default = "us-east-1"
    }

    # Fetch OIDC provider thumbprint for root CA
    # Documentation: https://www.terraform.io/docs/language/data-sources/index.html
    data "external" "changeme_external_thumbprint_data" {
      program    = ["\${path.module}/oidc_thumbprint.sh", var.changeme_spot_and_fargate_region]
      depends_on = [aws_eks_cluster.changeme_spot_and_fargate_eks_cluster]
    }


    # Documentation: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_openid_connect_provider
    resource "aws_iam_openid_connect_provider" "changeme_iam_openid_connect_provider" {
      client_id_list  = ["sts.amazonaws.com"]
      thumbprint_list = [data.external.changeme_external_thumbprint_data.result.thumbprint]
      url             = data.aws_eks_cluster.changeme_spot_and_fargate_eks_cluster_data.identity[0].oidc[0].issuer

      lifecycle {
        ignore_changes = [thumbprint_list]
      }
    }
      `,
    [binding.aws, binding.external],
    Synth.yes,
    {
      dataSources: ["aws_eks_cluster", "external"],
      resources: ["aws_iam_openid_connect_provider", "aws_eks_cluster"],
    }
  );
});
