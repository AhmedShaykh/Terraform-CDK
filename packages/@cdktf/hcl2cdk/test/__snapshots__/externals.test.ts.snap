// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`externals handle external data typescript snapshot 1`] = `
"import * as constructs from \\"constructs\\";
import * as cdktf from \\"cdktf\\";
/*Provider bindings are generated by running cdktf get.
See https://cdk.tf/provider-generation for more details.*/
import * as aws from \\"./.gen/providers/aws\\";
import * as external from \\"./.gen/providers/external\\";
class MyConvertedCode extends cdktf.TerraformStack {
  constructor(scope: constructs.Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new aws.provider.AwsProvider(this, \\"aws\\", {
      region: \\"us-east-1\\",
    });
    new external.provider.ExternalProvider(this, \\"external\\", {});
    const changemeSpotAndFargateRegion = new cdktf.TerraformVariable(
      this,
      \\"changeme_spot_and_fargate_region\\",
      {
        default: \\"us-east-1\\",
      }
    );
    const awsEksClusterChangemeSpotAndFargateEksCluster =
      new aws.eksCluster.EksCluster(
        this,
        \\"changeme_spot_and_fargate_eks_cluster\\",
        {
          name: \\"foo\\",
          roleArn:
            \\"arn:aws:iam::123456789012:role/eksctl-foo-cluster-ServiceRole-1A2B3C4D5E6F\\",
          vpcConfig: {
            subnetIds: [\\"subnet-12345678901234567\\"],
          },
        }
      );
    const dataAwsEksClusterChangemeSpotAndFargateEksClusterData =
      new aws.dataAwsEksCluster.DataAwsEksCluster(
        this,
        \\"changeme_spot_and_fargate_eks_cluster_data\\",
        {
          name: \\"foo\\",
        }
      );
    const dataExternalChangemeExternalThumbprintData =
      new external.DataExternal(this, \\"changeme_external_thumbprint_data\\", {
        dependsOn: [awsEksClusterChangemeSpotAndFargateEksCluster],
        program: [
          \\"\${path.module}/oidc_thumbprint.sh\\",
          changemeSpotAndFargateRegion.value,
        ],
      });
    new aws.iamOpenidConnectProvider.IamOpenidConnectProvider(
      this,
      \\"changeme_iam_openid_connect_provider\\",
      {
        clientIdList: [\\"sts.amazonaws.com\\"],
        lifecycle: {
          ignoreChanges: [thumbprintList],
        },
        thumbprintList: [
          dataExternalChangemeExternalThumbprintData.result.thumbprint,
        ],
        url:
          \\"\${\\" +
          dataAwsEksClusterChangemeSpotAndFargateEksClusterData.fqn +
          \\"}.identity[0].oidc[0].issuer\\",
      }
    );
  }
}
"
`;
